File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.js
const input = require('fs').readFileSync(0).toString().split('\n');
let idx = 0;
const [N, M] = input[idx++].split(' ').map(Number);
const campus = [];
let startX = 0, startY = 0;

for (let i = 0; i < N; i++) {
    const row = input[idx++].trim().split('');
    campus.push(row);
    for (let j = 0; j < M; j++) {
        if (row[j] === 'I') {
            startX = i;
            startY = j;
        }
    }
}

function bfs(x, y) {
    const queue = [[x, y]];
    const visited = Array.from({length: N}, () => new Array(M).fill(false));
    visited[x][y] = true;
    let front = 0;
    let count = 0;

    while (front < queue.length) {
        const [cx, cy] = queue[front++];
        for (const [dx, dy] of [[-1,0],[1,0],[0,-1],[0,1]]) {
            const nx = cx + dx;
            const ny = cy + dy;
            if (0 <= nx && nx < N && 0 <= ny && ny < M && campus[nx][ny] !== 'X' && !visited[nx][ny]) {
                visited[nx][ny] = true
                if (campus[nx][ny] === 'P') {
                    count++;
                }
                queue.push([nx, ny]);
            }
        }
    }
    return count;
}

const result = bfs(startX, startY);
console.log(result > 0 ? result : 'TT');

File: 백준/Silver/21736. 헌내기는 친구가 필요해/헌내기는 친구가 필요해.py
from collections import deque
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
campus = []
start = (0, 0)

for i in range(N):
    row = list(input().strip())
    campus.append(row)
    for j in range(M):
        if row[j] == 'I':
            start = (i, j)

def bfs(x, y):
    queue = deque([(x, y)])
    visited = [[False] * M for _ in range(N)]
    visited[x][y] = True
    count = 0
    
    while queue:
        cx, cy = queue.popleft()
        for dx, dy in [(-1,0),(1,0),(0,-1),(0,1)]:
            nx, ny = cx + dx, cy + dy
            if 0 <= nx < N and 0 <= ny < M and campus[nx][ny] != 'X' and not visited[nx][ny]:
                visited[nx][ny] = True
                if campus[nx][ny] == 'P':
                    count += 1
                queue.append((nx, ny))
    
    return count

result = bfs(*start)
print(result if result > 0 else 'TT')